Separate schema matching into read/write layers

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59465 Change `NoneStatus` from unscoped enum to scoped enum
* **#59406 Separate schema matching into read/write layers**
* #59245 Factor out `CompilationUnit` definitions into implementation file
* #59244 Move helper structs/methods out of `ir_emitter.cpp`

Differential Revision: [D28880522](https://our.internmc.facebook.com/intern/diff/D28880522)

Commit:
Separate schema matching into read/write layers

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28880522).

Commit:
Update on "Separate schema matching into read/write layers"


Differential Revision: [D28880522](https://our.internmc.facebook.com/intern/diff/D28880522)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/schema_emitter.cpp:
```diff
@@ -0,0 +1,144 @@
+namespace torch {
+namespace jit {
+
+// pack outputs of a function following python rules. If there is a single value
+// return a SimpleValue, otherwise pack all the values into a Tuple.
+static Value* packOutputs(
+    Graph& g,
+    at::ArrayRef<Value*> values,
+    c10::OptNameList field_names) {
+  if (values.size() == 1) {
+    return values[0];
+  }
+  std::shared_ptr<FunctionSchema> schema;
+  TupleTypePtr named_tuple = nullptr;
+  if (field_names) {
+    auto types = fmap(values, [](Value* v) { return v->type(); });
+    named_tuple =
+        TupleType::createNamed(c10::nullopt, field_names.value(), types);
+  }
+  return g.insertNode(g.createTuple(values, named_tuple))->output();
+}
+
```
Are `packOutputs`, `emitBuiltinNode` and `emitBuiltinCall` just moved or moved and modified? I didn't spot any differences, assuming they are not modified, thus not carefuly reviewed. Let me know if this is not the case.

ansley replied:
Just moved :)

gmagogsfm commented on torch/csrc/jit/frontend/schema_matching.cpp:
```diff
@@ -464,6 +476,7 @@ static c10::optional<MatchedSchema> tryMatchSchema(
   return MatchedSchema{
       std::move(positional_inputs),
       std::move(return_types),
+      {std::move(additions)},
```
OOC, why is the curly bracket needed here?

ansley replied:
Good catch--holdout from when I had a slightly different API

gmagogsfm replied:
Should the curly brackets be removed then?

gmagogsfm commented on torch/csrc/jit/frontend/schema_matching.h:
```diff
@@ -39,25 +41,22 @@ TORCH_API bool convertibleToList(
     const std::string& name,
     at::ArrayRef<NamedValue> kwargs);
 
 // applies implicit conversion from value trying to turn it into type
```
Do you mean to remove these two lines of old comment?

gmagogsfm replied:
Ping

ansley replied:
Thanks!

ansley commented on torch/csrc/jit/frontend/schema_emitter.cpp:
```diff
@@ -0,0 +1,144 @@
+#include <torch/csrc/jit/frontend/schema_emitter.h>
```
I agree. I had a `TODO` in `schema_emitter.h`, but that was a bit lazy...I've since been able to set everything up so that this logic can be in `ir_emitter_utils.h`

ansley commented on torch/csrc/jit/frontend/schema_matching.cpp:
```diff
@@ -326,6 +335,7 @@ static c10::optional<MatchedSchema> tryMatchSchema(
   TypeEnv type_env;
   std::vector<Value*> positional_inputs;
   std::vector<bool> used_kwarg(kwargs.size(), false);
+  auto additions = std::shared_ptr<Graph>(new Graph());
```
Fixed

Commit:
Update on "Separate schema matching into read/write layers"


Differential Revision: [D28880522](https://our.internmc.facebook.com/intern/diff/D28880522)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/resolver.h:
```diff
@@ -10,6 +10,9 @@ namespace jit {
 struct Resolver;
 using ResolverPtr = std::shared_ptr<Resolver>;
 
+// struct BuiltinModule;
+// struct FunctionValue;
```
Accident?

ansley replied:
Yes

Commit:
Update on "Separate schema matching into read/write layers"


Differential Revision: [D28880522](https://our.internmc.facebook.com/intern/diff/D28880522)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/schema_emitter.cpp:
```diff
@@ -0,0 +1,145 @@
+#include <torch/csrc/jit/frontend/schema_matching.h>
```
Nit:

I think this `schema_emitter.cpp` meant to match with `ir_emitter_utils.h`. It is customary to include its corresponding header file as first `include`. Also, I think it is better to change this file name to match `ir_emitter_utils.h`.

gmagogsfm reviewed (approved):
Looks good, just a few minor nits left.
